golangServer/api/import_by_url.go

```diff
@@ -0,0 +1,144 @@
+package api
+
+import (
+	"Oracle.com/golangServer/Oracle"
+	"Oracle.com/golangServer/config"
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+)
+
+func ImportEventListener() {
+	// Import channels for logs
+	Logs := make(chan *Oracle.OracleImportFromUrl)
+	// Subscribe to each event
+	opts := &bind.WatchOpts{Context: context.Background(), Start: nil}
+	eventSub, err := config.OracleContract.WatchImportFromUrl(opts, Logs)
+	if err != nil {
+		log.Fatal("Failed to subscribe to Import events:", err)
+	}
+	// start Listening...
+	log.Println("ImportEvent Listening ...")
+	for {
+		select {
+		case err := <-eventSub.Err():
+			log.Println("[Error in Event IMPORT:", err)
+		case event := <-Logs:
+			log.Println("Received put event ", event.ReqID)
+			importByUrl(event)
+		}
+	}
+}
+
+// Import data to memory db
+func importByUrl(event *Oracle.OracleImportFromUrl) {
+	var statement bool
+	tps := GenTransactOpts(config.GasLimit)
+
+	colName := event.ColName
+	ctx := context.Background()
+	dbName := event.DbName
+	db := config.Dbs[dbName]
+	dbC, err := db.Collection(colName, "")
+	if err != nil {
+		log.Println("Get collection ERROR: ", err)
+		info := fmt.Sprintf("Get collection ERROR: %v", err)
+		statement = false
+		//response to oracle
+		config.OracleContract.ImportFromUrlRsp(tps, event.ReqID, statement, event.Sender, info)
+		return
+	}
+
+	resp, err := http.Get(event.Url)
+	defer resp.Body.Close()
+	if err != nil {
+		log.Println("Get datas ERROR: ", err)
+		info := fmt.Sprintf("Get datas ERROR: %v", err)
+		statement = false
+		//response to oracle
+		config.OracleContract.ImportFromUrlRsp(tps, event.ReqID, statement, event.Sender, info)
+		return
+	}
+	if resp.StatusCode != 200 {
+		log.Println("Get datas Fail, StatusCode = ", resp.StatusCode)
+		info := fmt.Sprintf("Get datas Fail, StatusCode = %v", resp.StatusCode)
+		statement = false
+		//response to oracle
+		config.OracleContract.ImportFromUrlRsp(tps, event.ReqID, statement, event.Sender, info)
+		return
+	}
+	if resp.Header.Get("Content-Length") == "0" {
```

Comment:
We should inform the oracle the max number of bytes we can import and have it end an error if the file is too big. Or is this already being done elswhere?

Reply:
File size determination has now been added

Reply:
Sorry I think I explained poorly, I meant the max number of bytes that the user paid for:

- user starts import transaction
- they choose how many bytes they wish to pay for and pay that much eth
- oracle receives request and only imports the number of bytes that were paid for

Reply:
I think users should prepare the data to be imported in advance, rather than controlling the number of bytes of data imported by paying fees, as this may be an unsafe practice.